packages/financial-templates-lib/src/clients/OptimisticOracleClient.js

```diff
@@ -0,0 +1,118 @@
+   * @param {Object} logger Winston module used to send logs.
+   * @param {Object} oracleAbi OptimisticOracle truffle ABI object to create a contract instance.
+   * @param {Object} web3 Provider from Truffle instance to connect to Ethereum network.
+   * @param {String} oracleAddress Ethereum address of the OptimisticOracle contract deployed on the current network.
+   * @return None or throws an Error.
+   */
+  constructor(logger, oracleAbi, web3, oracleAddress) {
+    this.logger = logger;
+    this.web3 = web3;
+
+    // Oracle contract
+    this.oracle = new web3.eth.Contract(oracleAbi, oracleAddress);
+    this.oracleAddress = oracleAddress;
+
+    // Oracle Data structures & values to enable synchronous returns of the state seen by the client.
+    this.priceRequests = [];
+    this.priceProposals = [];
+
+    // Store the last on-chain time the clients were updated to inform price request information.
+    this.lastUpdateTimestamp = 0;
+
+    // Helper functions from web3.
+    this.hexToUtf8 = this.web3.utils.hexToUtf8;
+  }
+
+  // Returns an array of Price Request objects that have no proposals yet.
+  getAllPriceRequests() {
+    return this.priceRequests;
+  }
+
+  // Returns an array of Price Request objects that someone has proposed a price for.
+  getAllPriceProposals() {
+    return this.priceProposals;
+  }
+
+  // Returns an array of Price Request objects for each position that someone has proposed a price for and whose
+  // proposal can be disputed. The proposal can be disputed because the proposed price deviates from the `inputPrice` by
+  // more than the `errorThreshold`.
+  getDisputablePriceProposals() {
+    // TODO
+    return;
+  }
+
+  // Returns the last update timestamp.
+  getLastUpdateTime() {
+    return this.lastUpdateTimestamp;
+  }
+
+  async update() {
+    // Fetch contract state variables in parallel.
+    const [requestEvents, currentTime] = await Promise.all([
+      this.oracle.getPastEvents("RequestPrice", { fromBlock: 0 }), // TODO: Change this `fromBlock` to something > 0
+      this.oracle.methods.getCurrentTime().call()
+    ]);
+
+    // Fetch request states in parallel batches, 20 at a time, to be safe and not overload the web3 node.
+    const WEB3_CALLS_BATCH_SIZE = 150;
+
+    const [requestStates] = await Promise.all([
+      Promise.map(
+        requestEvents,
+        request =>
+          // Since we're making an async web3 call here, consider calling
+          // requests(web3.utils.soliditySha3(requester, id, timestamp, acData)) instead
+          // so we can grab more data including that relevant for proposals.
+          this.oracle.methods
+            .getState(
+              request.returnValues.requester,
+              request.returnValues.identifier,
+              request.returnValues.timestamp,
+              request.returnValues.ancillaryData ? request.returnValues.ancillaryData : "0x"
+            )
+            .call(),
```

Comment:
Hmmm, I'm curious about how to make this scalable. I know we're trying to rethink how much we're relying on big event queries, but this could very quickly blow through a lot of infura quota if I'm reading this correctly, since it seems that each run would query the state of every past request (unless you're planning to pre-filter the events in some way?). Thoughts?

Reply:
Hm yeah we need a caching solution once a bot has been running for a while. I think on startup the default START_BLOCK can be the current time, but then once its going it can store price request data in its state. WDYT

Reply:
We have to think about what states should we no longer track price requests, for example once a PriceRequest is `REQUESTED` and responded to, then it should no longer be checked in the `propose()` method, but once it is `PROPOSED` and either disputed/not-disputed, then it should no longer be checked in the `dispute()` method.

Reply:
As discussed IRL, IMO, I think it makes sense to look for any requests that have had "activity" within the last `n` days. We could get more specific, depending on the type of activity.

For instance, we could consider any request that:
- Has been requested within the last 1 day (arbitrary -- this could be as low as 2 hours IMO). I don't think there's any danger in "forgetting" about requested prices, since it will show up with a proposed event if someone proposes a price.
- Has been proposed within the last 2 hours. Even though the liveness can differ from the default, we generally expect that our bots will dispute bad proposals within 2 hours, so the only danger in a proposal being "forgotten" after 2 hours is if this bot is down for the entire 2 hours. Otherwise, it would dispute it.
- Has been disputed within the past 4 days (only if the bot intends to settle disputes).

The simplest solution, though, is probably just to search over the past 4 days or so (or 1 week for a more round number) for all of these event types.

Reply:
So it seems like one simple solution would be to parameterize a `lookback` that we can use to filter any requests that occurred (or had some other action) earlier than.

Reply:
Yeah, I think a parameterizable lookback with a sensible default is the right way to go. +1

Reply:
Done PTAL, I ended up replacing the `getFromBlock` (@pemulis 's implementation) with the block derived from the `lookbackWindow` parameter.